Zephyr cavs18

lyakh opened the pull request:
Zephyr fixes for cAVS 1.8+. Note, that the first commit has already been merged, the last one is actually a temporary "fix," but probably we'll have to apply it for now or find a different way to limit the number of cores to 1

lgirdwood commented on src/arch/xtensa/configs/cannonlake_gcc_defconfig:
```diff
@@ -3,6 +3,6 @@ CONFIG_INTEL_DMIC=y
 CONFIG_INTEL_ALH=y
 CONFIG_CAVS_LPS=y
-CONFIG_CORE_COUNT=4
+CONFIG_CORE_COUNT=1
```
Should we have an error/check here since today we can have CONFIG_MULTICORE=y and CORE_COUNT=1
it may be better to have a #error e.g.
```c
#ifdef __ZEPHYR__
#if CORE_COUNT > 1
#error blah blah blah
#endif 
#endif 
```
This ways its obvious core count must be 1 and it can be merged (and later reverted).

lyakh replied:
There are several configuration options, related to SMP support, used in SOF under Zephyr: `CONFIG_CORE_COUNT`, `CONFIG_SMP`, `CONFIG_MP_NUM_CPUS`, and then of course there's also `CONFIG_MULTICORE` in SOF itself. To make matters even worse they're "defined" in multiple places. E.g. in zephyr/kernel/Kconfig
```
config SMP
	bool "Enable symmetric multithreading support"
	depends on USE_SWITCH
	help
	  When true, kernel will be built with SMP support, allowing
	  more than one CPU to schedule Zephyr tasks at a time.

config MP_NUM_CPUS
	int "Number of CPUs/cores"
	default 1
	range 1 4
	help
	  Number of multiprocessing-capable cores available to the
	  multicpu API and SMP features.
```

and then, e.g. in zephyr/soc/xtensa/intel_adsp/cavs_v18/Kconfig.defconfig.series
```
config SMP
       default y

config MP_NUM_CPUS
	default 2
```

I tried to average out all those options a bit, the result is in https://github.com/andyross/zephyr/pull/1/commits/4d7f6ab34ce537ad9b77adab2b164293dd86fdeb So, I'm not really sure which config option is the "top level" multi-core support that we should be checking for?

lgirdwood replied:
ok, these need aligned to use the same as Zephyr. Can we have a mergable build error for now and we can fix properly with @andyross for Zephyr 2.5 and SOF v1.7

lgirdwood reviewed (changes requested):
Can you also remove the patches already in master

lyakh commented:
@lgirdwood this version should be better. It leaves CONFIG_CORE_COUNT alone and only modifies PLATFORM_CORE_COUNT to reflect the current temporary Zephyr restriction

lgirdwood commented on src/schedule/timer_domain.c:
```diff
@@ -42,24 +42,24 @@
 
 K_THREAD_STACK_DEFINE(ll_workq_stack0, ZEPHYR_LL_WORKQ_SIZE);
-#if CONFIG_CORE_COUNT > 1
+#if PLATFORM_CORE_COUNT > 1
```
Hmm, why do we have CONFIG_CORE_COUNT and PLATFORM_CORE_COUNT ? They look like the same thing, we should only have one. Could this fix/merge be done as a separate fix as it's confusing and should come from Kconfig.

lyakh replied:
I found a way to achieve the same without using `PLATFORM_CORE_COUNT`, I'll make a separate patch to get rid of it

lgirdwood reviewed (commented):
Looks good - I think we need to sore out the core count as well though.

lyakh commented:
@lgirdwood sorry, please wait with this one, there seems to be a problem I've overlooked

lyakh commented:
> @lgirdwood sorry, please wait with this one, there seems to be a problem I've overlooked

@lgirdwood it doesn't seem to work the way I tried to force the number of cores to 1 for Zephyr while keeping it unchanged for SOF stand-alone. At the moment we have `CONFIG_CORE_COUNT=1` in src/arch/xtensa/configs/cannonlake_defconfig, I tried to override that for Zephyr by adding `range 1 1 if KERNEL_BIN_NAME = "zephyr"` to `config CORE_COUNT` in src/arch/xtensa/Kconfig but this fails during configuration merging (no idea how I tested it last week. Everything works better on Friday (TM)).
We could for now keep the `PLATFORM_CORE_COUNT` macro and set it to 1 instead of `CONFIG_CORE_COUNT` under Zephyr - like my original version did. I'm looking for other options too though.

lgirdwood commented:
@lyakh make sure either macro is not duplicated on zephyr as this could explain that it worked by accident.

lyakh commented:
@lgirdwood this has been adjusted to match #3515 and the checkpatch warning isn't added by this PR

lyakh commented:
This isn't for cAVS 1.8 any more, it's also needed for the primary initial target: Apollolake. I've dropped one patch - a check for SMP, for it to work a patch is needed for Zephyr too. With just this set it's possible to build working Zephyr-based firmware for Up^2, using https://github.com/andyross/zephyr/tree/sof-rebased-14Oct20
I added 4 reverts for recent commits. Firmware images built with them don't run because of a missing / wrong firmware ABI version. To re-enable those commits we have to enable respective firmware manifest sections in Zephyr builds.

lgirdwood reviewed (changes requested):
Can you remove the reverts, we can fix this in Zephyr later once CI is working.

lyakh commented:
outdated
